### PR TITLE
feat: generate config from Proto

### LINKED ISSUE #1521: Create a module to generate a tailcall configuration from existing GRPC endpoints
We need to create a module that takes in a set of protobuf files, identifies all the methods, and generates a Tailcall Configuration automatically. 

**Technical Requirements**
- Write tests to create a configuration from protobuf.
- Make sure documentation is copied into the generated GraphQL config.
- We don't need to integrate with the CLI just yet.

**Additional Information**
This is easy to do. Write a module that takes a `Vec<FileDescriptorSet>` and produces a `Config`.

### PR DESCRIPTION
**Pull request by ssddOnTop**
**Issue Reference(s):**  
Fixes #1521 
/claim 1521

**TODO**

- [ ] Support doc auto gen for gql 
- [x] A test for a type used as both input and output
- [x] A test for mutation
- [x] A test for empty objects
- [x] A test for two types having same name in different packages
- [x] A test for required type
- [x] A test for optional type
- [x] A test for enum
- [x] A test for repeated fields

## Summary by CodeRabbit

- **New Features**
	- Introduced support for Enum types in GraphQL schema definitions, enhancing the flexibility and expressiveness of schema design.
	- Added new protocol buffer definitions for various functionalities, including news entities, phone numbers, and greeting services, facilitating richer data modeling and service operations.
	- Implemented a new test for enums with default values in a gRPC service, ensuring robustness and correctness in enum handling.

- **Enhancements**
	- Improved the logic for handling type variants and enum types in schema generation, resulting in more accurate and efficient schema representations.
	- Updated the functionality for converting protocol buffer definitions into structured configurations, streamlining the process of working with protocol buffers.

- **Bug Fixes**
	- Fixed issues related to dangling resolvers in schemas, ensuring the integrity and reliability of GraphQL schema definitions.

- **Refactor**
	- Renamed and clarified the purpose of internal functions for better maintainability and understanding of the codebase.

- **Tests**
	- Introduced new protocol buffer files for testing purposes, covering a wider range of scenarios and ensuring comprehensive testing coverage.

### CODE REVIEW

File: src/config/from_proto.rs [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+        )?;
+        set.file.push(file_desc);
+
+        let result = from_proto(set, None)?;
+        let cfg = Config::from_sdl(result.to_sdl().as_str()).to_result()?;
+
+        assert_eq!(cfg, result);
+        Ok(())
+    }
+}
```
[the remark this answers is not in the document]
Comment (tusharmath) [reply src/config/from_proto.rs]: Suggest all scenarios where tests should be added. WE need a good solution that converts any gRPC file into a GraphQL file.

File: src/config/reader.rs [hunk 2]
```diff
@@ -261,8 +261,8 @@ impl ConfigReader {
 
         while let Some(file) = queue.pop_front() {
             for import in file.dependency.iter() {
-                let proto = self.read_proto(import).await?;
                 if descriptors.get(import).is_none() {
+                    let proto = self.read_proto(import).await?;
```
Comment (ssddOnTop) [inline src/config/reader.rs:265]: it used to unnecessarily read proto files.. it was just 1 line change so I kept it in this pr

File: src/config/proto_generator.rs [hunk 3]
```diff
@@ -0,0 +1,38 @@
+use std::sync::Arc;
+
+pub struct ProtoGeneratorConfig {
```
Comment (tusharmath) [inline src/config/proto_generator.rs:3]: Keep the derived setters. We will need it when we wish to configure.

File: src/grpc/tests/proto/news_enum.proto [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    string postImage = 4;
+    Status foo = 5;
+    message Author { // nested field
+        string name = 1;
+        string email = 2;
+    }
+    Author author = 6;
+}
+
+// my awesome doc for service
```
Comment (tusharmath) [inline src/grpc/tests/proto/news_enum.proto:25]: comment not required.

File: src/grpc/tests/proto/news_enum.proto [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+service NewsService {
+    rpc GetAllNews (google.protobuf.Empty) returns (NewsList) {}
+    rpc GetNews (NewsId) returns (News) {}
+    rpc GetMultipleNews (MultipleNewsId) returns (NewsList) {}
+    rpc DeleteNews (NewsId) returns (google.protobuf.Empty) {}
+    rpc EditNews (News) returns (News) {}
+    rpc AddNews (News) returns (News) {}
```
Comment (tusharmath) [inline src/grpc/tests/proto/news_enum.proto:32]: Need tests for 
- [ ] A test for a type used as both input and output
- [ ] A test for mutation
- [ ] A test for empty objects
- [ ] A test for two types having same name in different packages

**Comment by ssddOnTop** [comment]
closing in favor of #1552

**Pull request closed without merging**